### PR TITLE
HDDS-8707. Avoid linear search in DBDefinition implementations.

### PR DESCRIPTION
**Pull request by szetszwo**
## What changes were proposed in this pull request?

DBDefinition declares the following method returning an array.
```java
public DBColumnFamilyDefinition[] getColumnFamilies();
```
It forces the callers to perform a linear search for looking up DBColumnFamilyDefinition by names.  The method should return a `Map`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8707

## How was this patch tested?

By existing unit tests.

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @szetszwo for the improvement.  LGTM, except an edge case in the multi-map iterator, which does not affect its current usage with the DB definitions.  I can post a follow-up PR for that if you prefer.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/CollectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      public T next() {
+        if (hasNextItem()) {
+          return i.next();
+        }
+        if (listIterator.hasNext()) {
+          i = listIterator.next().iterator();
+          return i.next();
+        }
+        throw new NoSuchElementException();
+      }
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/CollectionUtils.java:84]: @adoroszlai , good catch!  You are right that it won't work if there are empty lists.   Thanks for providing the code.  Let me try it.

**Comment by szetszwo** [comment]
When merging this, please add 
```
Co-authored-by: Doroszlai, Attila <adoroszlai@apache.org>
```

**Comment by adoroszlai** [comment]
Thanks @szetszwo for updating the patch.

**Pull request merged**